CuVS-Java: Update readme files (#672)
rapidsai/cuvs: java/README.md

# CuVS Java API

> CuVS Java API attempts to provide a Java based simple, efficient, and a robust vector search API.

Reviewer: ```suggestion
CuVS Java API provides a Java based simple, efficient, and a robust vector search API.
```

We don't attempt here, we just do.

> [!CAUTION]

> > This is an experimental release and updates to this API should be expected in the coming days.

Reviewer: I've learned that timeboxing like this is not a good idea. We *should* be clear that APIs are experimental, but they'll be this way for the 25.02 release. Great if we can remove them in the 25.04 release, but this stable release will have experimental APIs. 

```suggestion
> cuVS 25.02 contains an experimental version and updates to this API are expected in the coming release.
```

## Prerequisites

> - gcc 11.4

Reviewer: Is this the result of cuVS being dependent upong gcc 11.4? Is gcc 11.4 really a requirement or can users still use GCC 9.x (which is the requirement for buildign cuVS). Let's link to the cuVS build from source page here if these dependencies are only for building cuVS itself. Having these types of constraints duplicated across the codebas becomes hard to manage and maintain.

> - [nvcc 12.4 or above](https://developer.nvidia.com/cuda-downloads)

Reviewer: cuVS supports 11.x and box. Is there a reason the java API only supports 12.4? If not, let's just point to the cuVS `build from source` docs here and not duplicate these things.

> - [cmake 3.28 or above](https://cmake.org/download/)

Reviewer: Same here- I'm going to assume the only differences are the maven version and JDK. Why ubuntu 22.04? it doesn't look good to only support a very specific version of linux. I propose removing it, since users can get the toolchain they need from building the proper conda environment.

- [maven 3.9.6 or above](https://maven.apache.org/download.cgi)
- [JDK 22](https://openjdk.org/projects/jdk/22/)
- [Ubuntu 22.04](https://releases.ubuntu.com/jammy/)

## Building

> libcuvs libraries are needed for this API. If libcuvs libraries are already not built please do `./build.sh libcuvs java` in the top level directory to build this API.

Reviewer: I'm really not a fan of having this embedded in the java directory. I'd much prefer that this information is provided directly in the `build cuVS from source` docs and this README just links to that. It just simply makes things hard to find, maintain (and update) when this info is scattered around the codebase.

Let's keep it here for the moment while this API is in experimental state, but plan to move into the docs in 25.04 release cycle. 

```suggestion
The libcuvs C and C++ libraries are needed for this API. If libcuvs libraries have not been built and installed, use `./build.sh libcuvs java` in the top level directory to build this API.
```

Alternatively, if libcuvs libraries are already built and you just want to build this API, please
do `./build.sh java` in the top level directory or just do `./build.sh` in this directory.

> :warning: If you notice the tests failing please replace `mvn verify` with `mvn clean package` in the `build.sh` script found in this directory and try again. This should build the API (and skip the tests).

Reviewer: This seems like a `notice` to me, not a warning.

## Examples

> For easy understanding we have provided starter examples for CAGRA, HNSW, and Bruteforce and these can be found in the `examples` directory.

Reviewer: ```suggestion
A few starter examples of CAGRA, HNSW, and Bruteforce index are provided in the `examples` directory.
```

Reviewer: Please make note that we need to move these to the `examples` directory at the top-level of this repo before we remove the experimental designator. All of the other code examples in cuVS are in that directory, so we expect users can look in a single place for them.

## Javadocs

> To generate javadocs, in this directory (after building the API) please do:

Reviewer: Before we remove the experimental designator, we need to get these building in CI and deployed.
